cpp/memilio/io/binary_serializer.h

```diff
@@ -0,0 +1,360 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+#ifndef MIO_IO_BINARY_SERIALIZER_H
+#define MIO_IO_BINARY_SERIALIZER_H
+
+#include "memilio/io/io.h"
+#include "memilio/io/serializer_base.h"
+#include "memilio/utils/compiler_diagnostics.h"
+#include "memilio/utils/metaprogramming.h"
+#include <memory>
+#include <sstream>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace mio
+{
+
+/**
+* In-memory stream of bytes.
+*/
+class ByteStream
+{
+    std::vector<unsigned char> m_buf;
+    size_t m_read_head;
+
+public:
+    /**
+    * Create stream with n readable bytes initialized to 0.
+    * Set value of readable bytes by accessing buffer through member function data().
+    * @param n number of readable bytes.
+    */
+    ByteStream(size_t n)
+        : m_buf(n, 0)
+        , m_read_head(0)
+    {
+    }
+
+    /**
+    * Create empty stream.
+    */
+    ByteStream()
+        : ByteStream(0)
+    {
+    }
+
+    /**
+    * Write bytes to streams.
+    * @param p pointer to bytes to be written.
+    * @param s number of bytes to write.
+    */
+    void write(const unsigned char* const p, size_t s)
+    {
+        m_buf.insert(m_buf.end(), p, p + s);
+    }
+
+    /**
+    * Read bytes from the stream.
+    * @param p pointer to buffer to read bytes into.
+    * @param s number of bytes to read.
+    * @returns true if bytes can be read, false if there are not enough readable bytes in the stream.
+    */
+    bool read(unsigned char* p, size_t s)
+    {
+        if (s <= (m_buf.size() - m_read_head)) {
+            auto read_begin = m_buf.begin() + m_read_head;
+            auto read_end   = read_begin + s;
+            std::copy(read_begin, read_end, p);
+            m_read_head += s;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+    * Reset the stream to n readable bytes of value 0.
+    * @param n number of readable bytes.
+    */
+    void reset(size_t n)
+    {
+        m_buf.assign(n, 0);
+        m_read_head = 0;
+    }
+
+    /**
+    * Reset the stream to empty.
+    */
+    void reset()
+    {
+        reset(0);
+    }
+
+    /**
+    * Get the pointer to the buffer of the stream.
+    * @{
+    */
+    const unsigned char* data() const
+    {
+        return m_buf.data();
+    }
+    unsigned char* data()
+    {
+        return m_buf.data();
+    }
+    /**@}*/
+
+    /**
+    * Get the size of the buffer of the stream.
+    */
+    size_t data_size() const
+    {
+        return m_buf.size();
+    }
+};
+
+/**
+* Stores a binary serialized object.
+* See io.h for documentation of serialization.
+*/
+class BinarySerializerObject : public SerializerBase
+{
+public:
+    BinarySerializerObject(std::shared_ptr<IOStatus> status, int flags, ByteStream& stream)
+        : SerializerBase(status, flags)
+        , m_stream(stream)
```

Comment:
json serializer has constructors with status-value-flags, status-flags, and value-status-flags.
binary serializer has status-flag-stream. 
It's consistently different every time, I don't see the problem!

Reply:
The pointer is passed by value, no need to make it const. The IOStatus object that the pointer points to can't be const since the serializer needs to modify it.